### DISCUSSION TITLE
Consumer Group Coordinator와 Consumer Group Leader의 정확한 차이가 무엇인가?

### DISCUSSION THREAD

**Discussion by yjinheon** [kyungjunleeme/kafka_study#6] [Q&A] [answered]
Consumer Group Coordinator는 브로커라고 보면 되는건가? Group Coordinator, Consumer Group Leader 각각의 역할은 무엇인가?

관련 영상 출처: https://www.youtube.com/watch?v=OxMdru93E6k&t=1209s
관련 KIP: https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Client-side+Assignment+Proposal

**Comment by kyungjunleeme** [comment]
저는 우선 주말에는  Consumer Group Coordinator가 브로커가 아니라고 생각했는데, 
https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Client-side+Assignment+Proposal

해당 내용을 다시 읽어보니 broker 이긴 하나 Consumer Group 안의 Consumer 들의 partition 할당에 관여하는 부분을 sever-side -> client-side로 옮긴것이 핵심 부분이기 때문에 -> 그렇기 때문에 server를 재시작하지 않아도 된다. 이게 핵심이기 때문에

이게 뭐 broker냐? consumer 영역이냐?는 중요한것은 아니라고 보이고, 약간 본질을 잃게 만든다고 생각이 들고
- 사실기준으로 말씀드리면
- Consumer Coordinator는 Consumer 쪽 내부 구현 쪽에 코드가 존재하고 있습니다. 

즉) 
- 하나의 broker가 Consumer Group Coordinator 역할을 해주는 것 뿐임. 우선 질문에 대한 답변으로는 broker가 맞음
- Consumer Group Coordinator는 -> Consumer Group Leader가 follower Consumer 간의 통신을 할 수 있는 역할을 해주는 Broker 지만 Client-side 쪽의 역할을 담당하는 녀석  그래서 하는 Cosumer쪽에 코드 위치

**Comment by kyungjunleeme** [comment, marked as answer]
https://github.com/a0x8o/kafka/blob/master/core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala 에 따르면 그룹코디네이터의 동작은 브로커에서 동작하는 것으로 보이고, 컨슈머 그룹 리더는 별도의 컨슈머가 해당 역할을 하는 것으로 알고 있습니다.

(인용 원영님 링크드인 답변)

**Comment by kyungjunleeme** [comment]
![image](https://github.com/user-attachments/assets/db9e30d0-82f8-405a-a87f-d45c768ba27b)
